src/server/detail/wrapped_json_path.h

```diff
@@ -44,16 +100,37 @@ template <typename T> class JsonCallbackResult {
   }
 
   JsonV1Result& AsV1() {
+    DCHECK(!ErrorOccured());
     return std::get<JsonV1Result>(result_);
   }
 
   JsonV2Result& AsV2() {
+    DCHECK(!ErrorOccured());
+    return std::get<JsonV2Result>(result_);
+  }
+
+  const JsonV1Result& AsV1() const {
+    DCHECK(!ErrorOccured());
+    return std::get<JsonV1Result>(result_);
+  }
+
+  const JsonV2Result& AsV2() const {
+    DCHECK(!ErrorOccured());
     return std::get<JsonV2Result>(result_);
   }
 
+  bool ErrorOccured() const {
+    return static_cast<bool>(error_code_);
+  }
+
+  const std::error_code& GetError() const {
+    return error_code_;
+  }
+
  private:
   std::variant<JsonV1Result, JsonV2Result> result_;
   bool legacy_mode_is_enabled_;
+  std::error_code error_code_;
```

Comment:
Maybe you don't need the bool variable, you set the variant based on it. Also it's a bit unusual to see error embedding (though we have it in places), if we can use Result<> (though it's not the most convenient to use)